homeassistant/components/epson/__init__.py

```diff
@@ -3,32 +3,35 @@
 import logging
 
 from epson_projector import Projector
-from epson_projector.const import POWER, STATE_UNAVAILABLE as EPSON_STATE_UNAVAILABLE
+from epson_projector.const import (
+    PWR_OFF_STATE,
+    STATE_UNAVAILABLE as EPSON_STATE_UNAVAILABLE,
+)
 
 from homeassistant.components.media_player import DOMAIN as MEDIA_PLAYER_PLATFORM
 from homeassistant.config_entries import ConfigEntry
-from homeassistant.const import CONF_HOST, CONF_PORT
+from homeassistant.const import CONF_HOST
 from homeassistant.core import HomeAssistant
-from homeassistant.helpers.aiohttp_client import async_get_clientsession
 
 from .const import DOMAIN
-from .exceptions import CannotConnect
+from .exceptions import CannotConnect, PoweredOff
 
 PLATFORMS = [MEDIA_PLAYER_PLATFORM]
 
 _LOGGER = logging.getLogger(__name__)
 
 
-async def validate_projector(hass: HomeAssistant, host, port):
+async def validate_projector(hass: HomeAssistant, host, check_powered_on=True):
     """Validate the given host and port allows us to connect."""
-    epson_proj = Projector(
-        host=host,
-        websession=async_get_clientsession(hass, verify_ssl=False),
-        port=port,
-    )
-    _power = await epson_proj.get_property(POWER)
-    if not _power or _power == EPSON_STATE_UNAVAILABLE:
-        raise CannotConnect
+    epson_proj = Projector(host=host, loop=hass.loop, type="tcp")
```

Comment:
Home Assistant requires to fetch some unique id from device. Epson allows to fetch serial number of it only through TCP.
so to work with config flow I can't allow http protocol anymore.

Reply:
Can you point to some docs? According to the developer docs, the unique ID is not an absolute requirement to make config-flow work.

https://developers.home-assistant.io/docs/config_entries_config_flow_handler

If there's a way how to keep also http (and other protocols over time), we should try to make it work.

Reply:
I don't see it in docs as well. I just see that if I don't set unique id then new entity is created every restart for many people.
one of issues: https://github.com/home-assistant/core/issues/45490

Reply:
Maybe there's a way how we can work it out. I suggest to wait for the core developers. Then, we get some more insight.

If it was an absolute necessity for config flow, we still have one option - to switch config flow schema to *tcp* and allow both protocols using `configuration.yaml`.

Thanks for referencing the issue, I will look at it.

Reply:
It seems very reasonable to use the projector's  serial number as a unique id.
@MartinHjelmare what do you thing?

Reply:
The config entry unique_id and the entity unique_id are two different concepts in Home Assistant. Some integrations set these to the same id, but that's integration specific.

https://developers.home-assistant.io/docs/config_entries_config_flow_handler#unique-ids

https://developers.home-assistant.io/docs/entity_registry_index#unique-id

It is not a requirement to set a unique_id of the config entry in the config flow. We do like to avoid setting up the same entry more than once so if it's not possible to set a unique_id in the config flow we deduplicate flows with some less specific information eg host ip without using the unique_id api.

A serial number is a great way to set a unique_id.

How will existing users be limited if we remove the http protocol? Are there different network criteria for http vs tcp which could affect users depending on their network environment?

According to the user documentation we do require users to connect the device to the LAN.

https://www.home-assistant.io/integrations/epson/

Reply:
The current problem is that we don't deduplicate a new config flow from existing config entries. It could be enough to just compare the flow host ip with existing entries and abort the flow if there are any matches. This is a common approach when a unique_id is not available, as I wrote above.

Reply:
> The config entry unique_id and the entity unique_id are two different concepts in Home Assistant. Some integrations set these to the same id, but that's integration specific.
> 
> https://developers.home-assistant.io/docs/config_entries_config_flow_handler#unique-ids
> 
> https://developers.home-assistant.io/docs/entity_registry_index#unique-id
> 
> It is not a requirement to set a unique_id of the config entry in the config flow. We do like to avoid setting up the same entry more than once so if it's not possible to set a unique_id in the config flow we deduplicate flows with some less specific information eg host ip without using the unique_id api.
> 
> A serial number is a great way to set a unique_id.

according to this review - https://github.com/home-assistant/core/pull/39697#discussion_r514934021 I haven't set any unique_id and for some reason for many users (not me!) on every restart epson integration created new config_entry.
So there might be my bug in implementation or some bug in this uniqueness of auto creating id.
Easiest way would be to fetch serial number, but it is only possible when device is turned on.

> 
> How will existing users be limited if we remove the http protocol? Are there different network criteria for http vs tcp which could affect users depending on their network environment?

HTTP is used in newer projectors only and it is unofficial.
TCP and Serial are using more official protocol ESC/VP.NET and older projectors are using this as well https://download.epson-europe.com/pub/download/3254/epson325483eu.pdf

I had one feedback week ago on HTTP vs TCP, volume down/up is not working via TCP for model TW650 and it is working for HTTP.
I think that HTTP commands are same for each projector and TCP commands vary between different models.
Maybe best would be to use TCP connection for config_flow (serial number can be fetched only via TCP connection) and HTTP for media_player entity.
I'm open to any suggestions.

TCP is using different port, what other network criteria are you thinking about?

> 
> According to the user documentation we do require users to connect the device to the LAN.
> 
> https://www.home-assistant.io/integrations/epson/

This is because many people would like to connect projector to RS232 and use serial connection to HA.
My projector doesn't have RS232 or serial USB so I cannot implement it and I wanted to emphasize it in docs, so nobody would try with `/dev/ttyUSB0` etc.

Reply:
> So there might be my bug in implementation or some bug in this uniqueness of auto creating id.
Easiest way would be to fetch serial number, but it is only possible when device is turned on.

The problem is, as I wrote above, that we don't deduplicate the host address of new flows compared to existing config entries. That would be an easy fix to add, that doesn't require any large changes.

We need to do this here, before validating the projector:
https://github.com/home-assistant/core/blob/fe1e57e76fd419a5eaed25ab6a779b2794984025/homeassistant/components/epson/config_flow.py#L33-L43

Reply:
Deduplicating on host address when the device actually has a true unique ID doesn't seem ideal though.
Maybe add a unique id to an old config entry be added once the device is online and matches on host as migration, and then remove that in a quarter or so?

Reply:
so as I understand all this there are 2 ways to fix:
- revert to HTTP and check if host address already exists in existing config, there won't be serial number as unique_id
- switch at least for initial config_flow to TCP, require from user to turn on projector during initialization, there will be unique id and for turn on/off and source select there will be a lot more models supported.

Reply:
It's common to require the device to be on during the config flow, but allow it to be off later when setting up the integration and have some mechanism of discovering the device when it turns on.

Having TCP in the config flow and http in the rest of the integration sounds ok if that will give the greatest compatibility with devices. It will be a bit more maintenance of course.

As codeowner I think you should decide which path to take. You know the device best.

Reply:
I used mixed protocol version. For config yaml I decided to compare host if not exists already in config yaml as I don't want to make users to turn on projector on each restart of HA.